app-cpp/appPtr.hpp

```diff
@@ -0,0 +1,481 @@
+#ifndef APP_PTR_H
+#define APP_PTR_H
+
+// std
+#include <atomic>
+#include <utility>
+
+namespace app355 {
+
+#pragma region unique_ptr
+
+/// @brief std::unique_ptr is a smart pointer that owns and manages a single object through 
+/// a pointer and disposes of that object when the unique_ptr goes out of scope.
+template<typename T>
+class unique_ptr
+{
+public:
+    unique_ptr() = default;
+    unique_ptr(T* inData);
+    ~unique_ptr();
+
+    void reset();
+    T* release();
+    T* get();
+
+    unique_ptr(unique_ptr&& ioMove);
+    unique_ptr& operator=(unique_ptr&& ioMove);
+    unique_ptr(unique_ptr& inCopy) = delete;
+    unique_ptr& operator=(unique_ptr& inCopy) = delete;
+
+    T* operator->();
+    T& operator*();
+    operator bool();
+
+private:
+    T* mData{};
+}; // class unique_ptr<T>
+
+// Inlines
+
+template<typename T>
+inline unique_ptr<T>::unique_ptr(T* inData) : 
+    //Beware: ControlBlock default initializes mStrongCount to 1
+    mData{inData}
+{
+    // Do Nothing
+}
+
+template<typename T>
+inline unique_ptr<T>::~unique_ptr()
+{
+    reset();
+}
+
+template<typename T>
+inline void unique_ptr<T>::reset()
+{
+    // delete is smart enough to do nothing when passed nullptr
+    delete mData;
+    mData = nullptr;
+}
+
+template<typename T>
+inline T* unique_ptr<T>::release()
+{
+    T* data = mData;
+    mData = nullptr;
+    return data;
+}
+
+template<typename T>
+inline unique_ptr<T>::unique_ptr(unique_ptr&& ioMove)
+{
+    reset();
+    mData = ioMove.release();
+}
+
+template<typename T>
+inline unique_ptr<T>& unique_ptr<T>::operator=(unique_ptr&& ioMove)
+{
+    // Moving to self should be a NOP
+    if (this != &ioMove)
+    {
+        reset();
+        mData = ioMove.release();
+    }
+    return *this;
+}
+
+template<typename T>
+inline T* unique_ptr<T>::operator->()
+{
+    return get();
+}
+
+template<typename T>
+inline T& unique_ptr<T>::operator*()
+{
+    return *get();
+}
+
+template<class T> 
+inline unique_ptr<T>::operator bool()
+{
+    return (mData != nullptr);
+}
+
+template<typename T>
+inline T* unique_ptr<T>::get()
+{
+    return mData;
+}
+
+// TRICKY: Variatic template arguments 
```

Comment:
Whenever you use `TRICKY`, also append this info to it...
`// TRICKY mnfitz 07apr2024: <blah, blah, blah>`

This is necessary info when you have software teams larger than 2 people.
So future maintainer knows how long this has been in code, and who to talk to.